File: attack-dataset/backend/orchestrator/engagement-store.js
/**
 * PostgreSQL engagement store — full document in payload JSONB.
 */

const { Pool } = require("pg");

const POSTGRES_DSN =
  process.env.POSTGRES_DSN || "postgresql://opsec:opsec@localhost:5432/attack_db";

const SCHEMA_SQL = `
CREATE TABLE IF NOT EXISTS engagements (
    id VARCHAR(255) PRIMARY KEY,
    target VARCHAR(255) NOT NULL,
    status VARCHAR(50) NOT NULL DEFAULT 'pending',
    source VARCHAR(64),
    payload JSONB NOT NULL DEFAULT '{}'::jsonb,
    started_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
    created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
    updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
);

CREATE INDEX IF NOT EXISTS idx_engagements_target ON engagements(target);
CREATE INDEX IF NOT EXISTS idx_engagements_status ON engagements(status);
CREATE INDEX IF NOT EXISTS idx_engagements_source ON engagements(source);
CREATE INDEX IF NOT EXISTS idx_engagements_started_at ON engagements(started_at DESC);
CREATE INDEX IF NOT EXISTS idx_engagements_updated_at ON engagements(updated_at DESC);

CREATE OR REPLACE FUNCTION update_engagements_updated_at()
RETURNS TRIGGER AS $$
BEGIN
    NEW.updated_at = NOW();
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS trg_engagements_updated_at ON engagements;
CREATE TRIGGER trg_engagements_updated_at
    BEFORE UPDATE ON engagements
    FOR EACH ROW EXECUTE FUNCTION update_engagements_updated_at();
`;

const LEGACY_MIGRATION_SQL = `
ALTER TABLE engagements ADD COLUMN IF NOT EXISTS source VARCHAR(64);
ALTER TABLE engagements ADD COLUMN IF NOT EXISTS payload JSONB DEFAULT '{}'::jsonb;
`;

function parsePayload(row) {
  if (!row) return null;
  let payload = row.payload;
  if (typeof payload === "string") {
    try {
      payload = JSON.parse(payload);
    } catch {
      payload = {};
    }
  }
  if (!payload || typeof payload !== "object") payload = {};
  const engagement = { ...payload, id: row.id || payload.id };
  if (!engagement.target) engagement.target = row.target;
  if (!engagement.status) engagement.status = row.status;
  if (!engagement.source && row.source) engagement.source = row.source;
  if (!engagement.started_at && row.started_at) {
    engagement.started_at =
      row.started_at instanceof Date ? row.started_at.toISOString() : row.started_at;
  }
  return engagement;
}

function extractIndexFields(engagement) {
  const id = String(engagement.id || "").trim();
  const target = String(engagement.target || "unknown").trim() || "unknown";
  const status = String(engagement.status || "pending").trim() || "pending";
  const source = engagement.source ? String(engagement.source).slice(0, 64) : null;
  let startedAt = engagement.started_at ? new Date(engagement.started_at) : new Date();
  if (Number.isNaN(startedAt.getTime())) startedAt = new Date();
  return { id, target, status, source, startedAt };
}

class EngagementStore {
  constructor() {
    this.pool = new Pool({
      connectionString: POSTGRES_DSN,
      max: 20,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 5000,
    });
    this.ready = this._bootstrap();
  }

  async _bootstrap(retries = 12, delayMs = 2000) {
    for (let attempt = 1; attempt <= retries; attempt++) {
      try {
        const client = await this.pool.connect();
        try {
          await client.query(SCHEMA_SQL);
          try {
            await client.query(LEGACY_MIGRATION_SQL);
          } catch {
            /* non-fatal */
          }
          console.log("[engagement-store] Schema ready");
          return;
        } finally {
          client.release();
        }
      } catch (error) {
        if (attempt === retries) {
          console.error("[engagement-store] Schema bootstrap failed:", error.message);
          throw error;
        }
        console.warn(
          `[engagement-store] Postgres not ready (attempt ${attempt}/${retries}), retrying…`
        );
        await new Promise((r) => setTimeout(r, delayMs));
      }
    }
  }

  async upsertEngagement(engagement) {
    await this.ready;
    const doc = { ...engagement };
    const { id, target, status, source, startedAt } = extractIndexFields(doc);
    if (!id) throw new Error("engagement.id is required");

    doc.id = id;
    doc.target = target;
    doc.status = status;
    if (source) doc.source = source;

    const query = `
      INSERT INTO engagements (id, target, status, source, payload, started_at, updated_at)
      VALUES ($1, $2, $3, $4, $5::jsonb, $6, NOW())
      ON CONFLICT (id) DO UPDATE SET
        target = EXCLUDED.target,
        status = EXCLUDED.status,
        source = EXCLUDED.source,
        payload = EXCLUDED.payload,
        updated_at = NOW()
      RETURNING *
    `;

    const result = await this.pool.query(query, [
      id,
      target,
      status,
      source,
      JSON.stringify(doc),
      startedAt,
    ]);
    return parsePayload(result.rows[0]);
  }

  /** @deprecated use upsertEngagement */
  async createEngagement(engagement) {
    return this.upsertEngagement(engagement);
  }

  async getEngagement(engagementId) {
    await this.ready;
    const result = await this.pool.query("SELECT * FROM engagements WHERE id = $1", [
      engagementId,
    ]);
    if (result.rows.length === 0) return null;
    return parsePayload(result.rows[0]);
  }

  async updateEngagement(engagementId, updates) {
    const existing = await this.getEngagement(engagementId);
    if (!existing) return false;
    return Boolean(await this.upsertEngagement({ ...existing, ...updates, id: engagementId }));
  }

  async deleteEngagement(engagementId) {
    await this.ready;
    const result = await this.pool.query("DELETE FROM engagements WHERE id = $1", [
      engagementId,
    ]);
    return result.rowCount > 0;
  }

  async listEngagements(options = {}) {
    await this.ready;
    const { limit = 1000, offset = 0, status, target } = options;
    const conditions = [];
    const values = [];
    let paramCount = 0;

    if (status) {
      paramCount++;
      conditions.push(`status = $${paramCount}`);
      values.push(status);
    }

    if (target) {
      paramCount++;
      conditions.push(`target ILIKE $${paramCount}`);
      values.push(`%${target}%`);
    }

    const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    values.push(limit, offset);

    const query = `
      SELECT * FROM engagements
      ${whereClause}
      ORDER BY COALESCE(started_at, updated_at) DESC
      LIMIT $${paramCount + 1} OFFSET $${paramCount + 2}
    `;

    const result = await this.pool.query(query, values);
    return result.rows.map(parsePayload);
  }

  async addLogEntry(engagementId, logEntry) {
    const engagement = await this.getEngagement(engagementId);
    if (!engagement) return false;
    const logs = Array.isArray(engagement.log) ? engagement.log : [];
    logs.push(logEntry);
    await this.upsertEngagement({ ...engagement, log: logs });
    return true;
  }

  async updateStatus(engagementId, status) {
    const engagement = await this.getEngagement(engagementId);
    if (!engagement) return false;
    const updates = { status };
    if (status === "complete" || status === "completed" || status === "error") {
      updates.completed_at = new Date().toISOString();
    }
    await this.upsertEngagement({ ...engagement, ...updates });
    return true;
  }

  async getEngagementStats() {
    await this.ready;
    const result = await this.pool.query(`
      SELECT
        COUNT(*)::int AS total,
        COUNT(*) FILTER (WHERE status IN ('pending', 'starting'))::int AS pending,
        COUNT(*) FILTER (WHERE status = 'scanning')::int AS scanning,
        COUNT(*) FILTER (WHERE status = 'building_vectors')::int AS building_vectors,
        COUNT(*) FILTER (WHERE status IN ('complete', 'completed'))::int AS complete,
        COUNT(*) FILTER (WHERE status IN ('error', 'failed'))::int AS error
      FROM engagements
    `);
    return result.rows[0];
  }

  async cleanupOldEngagements(days = 30) {
    await this.ready;
    const result = await this.pool.query(
      `DELETE FROM engagements
       WHERE (payload->>'completed_at')::timestamptz < NOW() - ($1::text || ' days')::interval
          OR (started_at < NOW() - ($1::text || ' days')::interval AND status IN ('pending', 'starting'))`,
      [String(days)]
    );
    return result.rowCount;
  }

  async close() {
    await this.pool.end();
  }
}

module.exports = EngagementStore;

File: attack-dataset/backend/shared/engagement_store.py
"""
Engagement persistence store for PostgreSQL.

Provides CRUD operations for engagement data persistence.
"""
from __future__ import annotations

import json
import uuid
from typing import Optional, Dict, Any, List
from datetime import datetime
import psycopg2
from psycopg2.extras import RealDictCursor, Json

from engagement_schema import initialize_engagement_schema


class EngagementStore:
    """PostgreSQL-based engagement store."""
    
    def __init__(self, postgres_dsn: str):
        """Initialize the engagement store with PostgreSQL connection."""
        self.conn = psycopg2.connect(postgres_dsn)
        self.conn.autocommit = True
        
        # Initialize schema
        initialize_engagement_schema(self.conn)
    
    def create_engagement(self, engagement_data: Dict[str, Any]) -> str:
        """Create a new engagement."""
        engagement_id = engagement_data.get('id', str(uuid.uuid4()))
        
        with self.conn.cursor() as cur:
            cur.execute("""
                INSERT INTO engagements (
                    id, target, status, aggression_level, boundary_profile,
                    scan_session, attack_chains, opsec_reports, opsec_audit,
                    analysis_overseer, ai_summary, log, started_at, error
                ) VALUES (
                    %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s
                )
            """, (
                engagement_id,
                engagement_data.get('target'),
                engagement_data.get('status', 'pending'),
                engagement_data.get('aggression_level', 1),
                Json(engagement_data.get('boundary_profile')),
                Json(engagement_data.get('scan_session')),
                Json(engagement_data.get('attack_chains')),
                Json(engagement_data.get('opsec_reports')),
                Json(engagement_data.get('opsec_audit')),
                Json(engagement_data.get('analysis_overseer')),
                engagement_data.get('ai_summary'),
                Json(engagement_data.get('log', [])),
                engagement_data.get('started_at', datetime.utcnow()),
                engagement_data.get('error')
            ))
        
        return engagement_id
    
    def get_engagement(self, engagement_id: str) -> Optional[Dict[str, Any]]:
        """Get an engagement by ID."""
        with self.conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute("SELECT * FROM engagements WHERE id = %s", (engagement_id,))
            row = cur.fetchone()
            
            if not row:
                return None
            
            # Convert to dict and handle JSONB fields
            result = dict(row)
            for field in ['boundary_profile', 'scan_session', 'attack_chains', 
                         'opsec_reports', 'opsec_audit', 'analysis_overseer', 'log']:
                if result.get(field) is not None:
                    result[field] = dict(result[field]) if isinstance(result[field], dict) else result[field]
            
            return result
    
    def update_engagement(self, engagement_id: str, updates: Dict[str, Any]) -> bool:
        """Update an engagement."""
        if not updates:
            return False
        
        # Build dynamic UPDATE query
        set_clauses = []
        values = []
        
        for key, value in updates.items():
            if key in ['boundary_profile', 'scan_session', 'attack_chains', 
                      'opsec_reports', 'opsec_audit', 'analysis_overseer', 'log']:
                set_clauses.append(f"{key} = %s")
                values.append(Json(value))
            else:
                set_clauses.append(f"{key} = %s")
                values.append(value)
        
        if not set_clauses:
            return False
        
        values.append(engagement_id)
        
        query = f"""
            UPDATE engagements 
            SET {', '.join(set_clauses)} 
            WHERE id = %s
        """
        
        with self.conn.cursor() as cur:
            cur.execute(query, values)
            return cur.rowcount > 0
    
    def delete_engagement(self, engagement_id: str) -> bool:
        """Delete an engagement."""
        with self.conn.cursor() as cur:
            cur.execute("DELETE FROM engagements WHERE id = %s", (engagement_id,))
            return cur.rowcount > 0
    
    def list_engagements(
        self, 
        limit: int = 50, 
        offset: int = 0,
        status: Optional[str] = None,
        target: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """List engagements with optional filters."""
        conditions = []
        values = []
        param_count = 0
        
        if status:
            param_count += 1
            conditions.append(f"status = ${param_count}")
            values.append(status)
        
        if target:
            param_count += 1
            conditions.append(f"target ILIKE ${param_count}")
            values.append(f"%{target}%")
        
        where_clause = ""
        if conditions:
            where_clause = "WHERE " + " AND ".join(conditions)
        
        values.extend([limit, offset])
        
        query = f"""
            SELECT * FROM engagements 
            {where_clause}
            ORDER BY started_at DESC 
            LIMIT %s OFFSET %s
        """
        
        with self.conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(query, values)
            rows = cur.fetchall()
            
            results = []
            for row in rows:
                result = dict(row)
                for field in ['boundary_profile', 'scan_session', 'attack_chains', 
                             'opsec_reports', 'opsec_audit', 'analysis_overseer', 'log']:
                    if result.get(field) is not None:
                        result[field] = dict(result[field]) if isinstance(result[field], dict) else result[field]
                results.append(result)
            
            return results
    
    def add_log_entry(self, engagement_id: str, log_entry: Dict[str, Any]) -> bool:
        """Add a log entry to an engagement's log."""
        engagement = self.get_engagement(engagement_id)
        if not engagement:
            return False
        
        logs = engagement.get('log', [])
        logs.append(log_entry)
        
        return self.update_engagement(engagement_id, {'log': logs})
    
    def update_status(self, engagement_id: str, status: str) -> bool:
        """Update engagement status."""
        updates = {'status': status}
        if status in ['complete', 'error']:
            updates['completed_at'] = datetime.utcnow()
        
        return self.update_engagement(engagement_id, updates)
    
    def get_active_engagements(self) -> List[Dict[str, Any]]:
        """Get all active (non-completed, non-error) engagements."""
        return self.list_engagements(
            limit=100,
            status='pending'  # You might want to use multiple statuses
        )
    
    def cleanup_old_engagements(self, days: int = 30) -> int:
        """Delete engagements older than specified days."""
        with self.conn.cursor() as cur:
            cur.execute("""
                DELETE FROM engagements 
                WHERE completed_at < NOW() - INTERVAL '%s days'
                OR (started_at < NOW() - INTERVAL '%s days' AND status = 'pending')
            """, (days, days))
            return cur.rowcount
    
    def get_engagement_stats(self) -> Dict[str, Any]:
        """Get engagement statistics."""
        with self.conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute("""
                SELECT 
                    COUNT(*) as total,
                    COUNT(CASE WHEN status = 'pending' THEN 1 END) as pending,
                    COUNT(CASE WHEN status = 'scanning' THEN 1 END) as scanning,
                    COUNT(CASE WHEN status = 'building_vectors' THEN 1 END) as building_vectors,
                    COUNT(CASE WHEN status = 'complete' THEN 1 END) as complete,
                    COUNT(CASE WHEN status = 'error' THEN 1 END) as error,
                    AVG(EXTRACT(EPOCH FROM (completed_at - started_at))) as avg_duration_seconds
                FROM engagements
            """)
            return dict(cur.fetchone())
    
    def close(self):
        """Close the database connection."""
        if self.conn:
            self.conn.close()